### ISSUE TITLE
Zeitwerk will setup twice

### ISSUE THREAD

**Issue by david50407** [fxn/zeitwerk#38]
reproduce steps (with ruby 2.6.1):

Gemfile:
```
source 'https://rubygems.org'
ruby '~> 2.6'
gem 'zeitwerk', '~> 2.1.2'
```

lib/foo.rb:
```
require 'zeitwerk'
loader = Zeitwerk::Loader.for_gem
loader.setup

module Foo; end
```

Run `bundle exec ruby lib/foo.rb` and get errors below:
```
Traceback (most recent call last):
        15: from lib/foo.rb:5:in `<main>'
        14: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/kernel.rb:16:in `require'
        13: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/kernel.rb:16:in `require'
        12: from /tmp/foo/lib/foo.rb:2:in `<top (required)>'
        11: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:417:in `for_gem'
        10: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/registry.rb:85:in `loader_for_gem'
         9: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/registry.rb:85:in `tap'
         8: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/registry.rb:88:in `block in loader_for_gem'
         7: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:187:in `push_dir'
         6: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:678:in `raise_if_conflicting_directory'
         5: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:678:in `synchronize'
         4: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:679:in `block in raise_if_conflicting_directory'
         3: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:679:in `each'
         2: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:682:in `block (2 levels) in raise_if_conflicting_directory'
         1: from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:682:in `each'
/home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:685:in `block (3 levels) in raise_if_conflicting_directory': loader (Zeitwerk::Error)

#<Zeitwerk::Loader:0x00000000021473d0
 @autoloaded_dirs=[],
 @autoloads={},
 @eager_load_exclusions=#<Set: {}>,
 @eager_loaded=false,
 @ignored=#<Set: {}>,
 @ignored_paths=#<Set: {}>,
 @inflector=
  #<Zeitwerk::GemInflector:0x0000000002158860
   @version_file=
    "/tmp/foo/lib/foo/version.rb">,
 @initialized_at=2019-04-15 16:40:21 +0800,
 @lazy_subdirs={},
 @logger=nil,
 @mutex=#<Thread::Mutex:0x0000000002144158>,
 @mutex2=#<Thread::Mutex:0x0000000002144108>,
 @preloads=[],
 @reloading_enabled=false,
 @root_dirs={},
 @setup=false,
 @shadowed_files=#<Set: {}>,
 @tag="foo",
 @to_unload={}>


wants to manage directory /tmp/foo/lib, which is already managed by

#<Zeitwerk::Loader:0x0000000002155e80
 @autoloaded_dirs=[],
 @autoloads=
  {"/tmp/foo/lib/foo.rb"=>[Object, "Foo"]},
 @eager_load_exclusions=#<Set: {}>,
 @eager_loaded=false,
 @ignored=#<Set: {}>,
 @ignored_paths=#<Set: {}>,
 @inflector=
  #<Zeitwerk::GemInflector:0x000000000217b978
   @version_file="lib/foo/version.rb">,
 @initialized_at=2019-04-15 16:40:21 +0800,
 @lazy_subdirs={},
 @logger=nil,
 @mutex=#<Thread::Mutex:0x0000000002154c38>,
 @mutex2=#<Thread::Mutex:0x0000000002154b48>,
 @preloads=[],
 @reloading_enabled=false,
 @root_dirs={"/tmp/foo/lib"=>true},
 @setup=true,
 @shadowed_files=#<Set: {}>,
 @tag="foo",
 @to_unload={}>
```

It seems `lib/foo.rb` loaded again when `loader.setup` called and caused another loader setups, and this caused crashes

**Comment by fxn** [comment]
Good one!

There are some inception-like use cases in which the file where `setup` is executed also defines a constant for which `setup` set an autoload.

This works correctly if that file is required, but this example shows it does not if exectued directly by Ruby.

I'll dig into it, but I am flying to RubyKaigi soon, so don't know when I will be able to.

If the module definition is not using constants from the project, an easy workaround meanwhile is to move the `setup` call after the module definition.

**Comment by fxn** [comment]
So the basic problem here is that we have this situation:

```ruby
# lib/foo.rb
autoload :Foo, File.expand_path(__FILE__)

module Foo
end

p :EXECUTED
```

You'll see `:EXECUTED` twice because when Ruby is about to define a module in the `Foo` constant, it detects it has an autoload for it, and executes the autoload. That is, it executes the file (again).

This does not happen when the entry point is required because `require` detects the file is being `required` and the autoload becomes a no-op.

Addressing this generically is unclear to me, I think I'd need some particular use case. Why are you executing a file under `lib` that way?

**Comment by david50407** [comment]
The original case I found is described below, the reproduce code above is the minimal code.

I designed a program named "Blah", and puts all files inside `lib` to match gem's file structure, then call `ruby lib/blah.rb` to run the program:

```
/
|- lib/
|  |- blah.rb # <- Init autoload here
|  |- blah/
|     |- cli.rb # class Blah::Cli
|- Gemfile
```

`lib/blah.rb`
```
require 'zeitwerk'

loader = Zeitwerk::Loader.for_gem
loader.setup

puts "Zeitwerk setted"
Blah::Cli.run
```

`lib/blah/cli.rb`
```
class Blah::Cli
  def self.run
    puts 'hello'
  end
end
```

`Gemfile`
```
source 'https://rubygems.org'

gem 'zeitwerk'
```

Then execute:
```txt
$ bundle install
Using bundler 1.17.3
Using zeitwerk 2.1.2
Bundle complete! 1 Gemfile dependency, 2 gems now installed.                                                                                                       
Use `bundle info [gemname]` to see where a bundled gem is installed.
$ ruby lib/blah.rb
Zeitwerk setted
/home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:685:in `block (3 levels) in raise_if_conflicting_directory': loader (Zeitwerk::Error)                         
                                                                                                                                                                                         
#<Zeitwerk::Loader:0x000000000217c210                                                                                                                                                    
 @autoloaded_dirs=[],                                                                                                                                                                    
 @autoloads={},                                                                                                                                                                          
 @eager_load_exclusions=#<Set: {}>,                                                                                                                                                      
 @eager_loaded=false,                                                                                                                                                                    
 @ignored=#<Set: {}>,                                                                                                                                                                    
 @ignored_paths=#<Set: {}>,                                                                                                                                                              
 @inflector=                                                                                                                                                                             
  #<Zeitwerk::GemInflector:0x0000000002187ef8                                                                                                                                            
   @version_file="/tmp/xxx/lib/blah/version.rb">,                                                                                                                              
 @initialized_at=2019-04-19 00:31:40 +0800,                                                                                                                                              
 @lazy_subdirs={},                                                                                                                                                                       
 @logger=nil,                                                                                                                                                                            
 @mutex=#<Thread::Mutex:0x00000000021879a8>,                                                                                                                                             
 @mutex2=#<Thread::Mutex:0x0000000002187f70>,                                                                                                                                            
 @preloads=[],                                                                                                                                                                           
 @reloading_enabled=false,                                                                                                                                                               
 @root_dirs={},                                                                                                                                                                          
 @setup=false,                                                                                                                                                                           
 @shadowed_files=#<Set: {}>,                                                                                                                                                             
 @tag="blah",                                                                                                                                                                            
 @to_unload={}>      

wants to manage directory /tmp/xxx/lib, which is already managed by                                                                                                            
                                                                                                                                                                                         
#<Zeitwerk::Loader:0x000000000217e420                                                                                                                                                    
 @autoloaded_dirs=[],                                                                                                                                                                    
 @autoloads={"/xxx/lib/blah.rb"=>[Object, "Blah"]},                                                                                                                        
 @eager_load_exclusions=#<Set: {}>,                                                                                                                                                      
 @eager_loaded=false,                                                                                                                                                                    
 @ignored=#<Set: {}>,                                                                                                                                                                    
 @ignored_paths=#<Set: {}>,                                                                                                                                                              
 @inflector=                                                                                                                                                                             
  #<Zeitwerk::GemInflector:0x000000000217dd68                                                                                                                                            
   @version_file="lib/blah/version.rb">,                                                                                                                                                 
 @initialized_at=2019-04-19 00:31:40 +0800,                                                                                                                                              
 @lazy_subdirs={"Blah"=>["/tmp/xxx/lib/blah"]},                                                                                                                                
 @logger=nil,                                                                                                                                                                            
 @mutex=#<Thread::Mutex:0x000000000217def8>,                                                                                                                                             
 @mutex2=#<Thread::Mutex:0x000000000217ded0>,                                                                                                                                            
 @preloads=[],                                                                                                                                                                           
 @reloading_enabled=false,                                                                                                                                                               
 @root_dirs={"/home/davy/tmp/xxx/lib"=>true},                                                                                                                                            
 @setup=true,                                                                                                                                                                            
 @shadowed_files=#<Set: {}>,                                                                                                                                                             
 @tag="blah",                                                                                                                                                                            
 @to_unload={}>                                                                                                                                                                          
                                                                                                                                                                                         
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:682:in `each'                                                                                          
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:682:in `block (2 levels) in raise_if_conflicting_directory'                                            
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:679:in `each'                                                                                          
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:679:in `block in raise_if_conflicting_directory'                                                       
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:678:in `synchronize'                                                                                   
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:678:in `raise_if_conflicting_directory'                                                                
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:187:in `push_dir'                                                                                      
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/registry.rb:88:in `block in loader_for_gem'                                                                      
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/registry.rb:85:in `tap'                                                                                          
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/registry.rb:85:in `loader_for_gem'                                                                               
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/loader.rb:417:in `for_gem'                                                                                       
  from /tmp/xxx/lib/blah.rb:3:in `<top (required)>'                                                                                                                            
  from /home/davy/.rvm/rubies/ruby-2.6.1/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'                                                                    
  from /home/davy/.rvm/rubies/ruby-2.6.1/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:54:in `require'                                                                    
  from /home/davy/.rvm/gems/ruby-2.6.1/gems/zeitwerk-2.1.2/lib/zeitwerk/kernel.rb:16:in `require'                                                                                        
  from lib/blah.rb:8:in `<main>'                                   
```

And you are right, after I create a wrapper `bin/blah` and required `lib/blah.rb` then create `module Blah; end` in `lib/blah.rb`, everything works fine.

`bin/blah`
```
#!/usr/bin/env ruby
require_relative '../lib/blah'
Blah::Cli.run
```

BTW, if I didn't declare `module Blah; end` inside `lib/blah.rb`, it cannot find `Blah` const.
Is this because of `lib/blah.rb`, which is matched to autoload rule to `Blah`, is loaded, so that Zeitwerk won't create a module fotr `Blah`?

**Comment by david50407** [comment]
> You'll see :EXECUTED twice because when Ruby is about to define a module in the Foo constant, it detects it has an autoload for it, and executes the autoload. That is, it executes the file (again).
> This does not happen when the entry point is required because require detects the file is being required and the autoload becomes a no-op.

So if we make the file `lib/foo.rb` which is in the issue description flagged as required (or excluded?) in Zeitwerk's overrided `Kernel.require`, and the problem also can be solved?

**Comment by fxn** [comment]
Yeah.

So the idiomatic way to do this is to have a script in `bin` that uses you code in `lib`.

If you have `lib/foo.rb` and `lib` is set as a root directory, `foo.rb` should define `Foo`, otherwise you need to opt-out ignoring the file somehow. The specifics depend on your project structure, but it is doable.

By now, I don't plan to support scripts, going to close this. But if you need help setting up your project please do not hesitate to followup.

**Comment by askegg** [comment]
This was helpful. For those who come after me:

I moved my Zeitwerk loading code into an initializer.rb file and required it from the module definition file.

initializer.rb
```ruby
require "zeitwerk"
loader = Zeitwerk::Loader.for_gem(warn_on_extra_files: false)
loader.setup
```
my_awesome_module.rb
```ruby
require_relative "initializer"

module MyAwesomeModule
...
```

**Comment by fxn** [comment]
@askegg that setup does not seem correct to me. First of all, in what sense is that related to this thread?

**Comment by askegg** [comment]
Hi @fxn. ☺️ 

Perhaps some background, because I think my situation was a little unusual.

I came across a video by Sandi Metz called "[All the little things](https://youtu.be/8bZh5LMaSmE)" where she refactored a program called The Gilded Rose. I found a repository online that contained a [Ruby version of the code](https://github.com/amckinnell/Gilded-Rose-Ruby) and thought it would be a good learning exercise.

Early in the exercise [I created a file for each class](https://github.com/askegg/Gilded-Rose-Ruby/commit/4f62ba2128d6350986042a74ce39116022bbd7cb), and since I didn't like all the require statements I added Zeitwerk. All the tests pass, however I was getting a very similar issue to the one above when I ran the code from a terminal. I think one of the reason is that the example code I started with is all contained in the "lib" directory.

After I [put the initialization code into its own file](https://github.com/askegg/Gilded-Rose-Ruby/commit/4f62ba2128d6350986042a74ce39116022bbd7cb) (as described above) I no longer saw those errors.

I hope all that makes sense.

**Comment by askegg** [comment]
This repository does not appear to be a gem - I just went with  Zeitwerk::Loader.for_gem because it worked. I am fumbling my way through this and was battling errors between rspec and running the code from the terminal, so when I managed to kludge a solution together I stopped fiddling.

After some tinkering this morning the initializer.rb now contains:
```ruby
require "zeitwerk"
loader = Zeitwerk::Loader.new
loader.push_dir(File.dirname(__FILE__))
loader.setup
```
This also works, and it seems closer to a proper implementation of Zeitwerk.

I appreciate your assistance here @fxn, but don't want to take up any more of your valuable time. This clearly not an issue with Zeitwerk, just my ignorance.
